Cache multiple MessageListData objects.

ryanreh99 opened the pull request:
Fixes #15131.
WIP as the point 5 is remaining.

timabbott commented:
I merged the first couple refactoring commits as the series ending with e5779c1ca0d55ec9b0dbeb9c96c747f326cffe84.  Next time, I'd prefer that you have split those commits (E.g. having the move-function-between-files be their own commits) as that makes it a lot faster to review, but I was able to verify the changes as it is.  It's great to see your fixing subtle bugs in this code path as you go.

@showell @amanagr FYI as folks who've worked in this code path recently.

timabbott commented on static/js/mld_store.js:
```diff
@@ -0,0 +1,84 @@
+exports.MldStore = MldStore;
+
+window.stored_mld = stored_mld;
+window.mld_store = exports;
```
Let's do this as a proper module, rather than using the `window` hack that we're trying to purge from the codebase.  The code under `static/shared` has some good examples of modern modules.  @andersk FYI to help with this.

I'd also like to think about naming for this module; I think `message_list_data_cache.js` might be a good name?  The `cache` part specifies that the intent/structure is that we can discard any of these that isn't in use for rendering at any time (either to save memory or because we aren't sure to how to update it properly) without introducing correctness issues.

timabbott commented on static/js/mld_store.js:
```diff
@@ -0,0 +1,84 @@
+
+    add(mld, key) {
+        if (key === undefined) {
+            key = hash_util.operators_to_hash(mld.filter.operators());
```
I guess given this structure, we should have a plan for how to destroy these objects when they become invalid (basically the `change_narrow` code path) unless we're confident that we're updating data structures properly.

timabbott commented on frontend_tests/node_tests/message_store.js:
```diff
@@ -340,6 +346,15 @@ run_test("message_id_change", () => {
+        muting_enabled: false,
+        filter: new Filter(operators),
+    });
+    stored_mld.add(mld);
```
@showell can you help @ryanreh99 develop a good node test suite for this data structure?  Ideally, such a test suite would have the property that we can do things like "confirm that message editing results in the right data structure".

timabbott commented on static/js/message_events.js:
```diff
@@ -391,12 +422,9 @@ exports.update_messages = function update_messages(events) {
-        list.remove_and_rerender(message_ids);
-    }
+    stored_mld.forEach((mld) => mld.remove(message_ids));
+    all_message_lists().forEach((ml) => ml.remove_and_rerender(message_ids));
```
I don't understand the logic here -- we're doing a removal on all message lists first, and then doing remove-and-rerender afterwards?  I think that'll be fragile or broken if a message list is in both sets, as the check for whether rerendering is required might be missed.

Would it work better to have these operations in the opposite order?

ryanreh99 replied:
I reversed the order for consistency with the other changes.
But the above order wouldn't be a problem since `remove_and_rerender` does a force `render`.

timabbott commented on static/js/muting.js:
```diff
@@ -53,6 +53,28 @@ exports.get_muted_topics = function () {
+
+    const remaining_muted_topics = exports.get_muted_topics();
+    return remaining_muted_topics.map((obj) => [obj.stream, obj.topic]);
+};
```
Why do we need this function?

ryanreh99 replied:
The is for real time sync. If a user unmutes an already muted topic from another tab,
the original tab has no way of knowing which MLD objects needs to get updated.
(As the event only contains all the muted topics)

timabbott commented:
This is really promising work @ryanreh99!  I'd love for us to do some focused work on writing tests that would allow us to be confident in the implementation's correctness before we extend to be actually used in part 5.  

In addition to unit tests, one idea would be that we do a simplified version of part 5 where we compare the MessageListData generated for rendering with the one fetched from this data structure and do a `blueslip.error` if they're materially different (which I guess we'd check by verifying that any differences are extra messages at the beginning or end?); that might provide an opportunity to test things out with real data on chat.zulip.org and gain confidence that the calculations are working correctly.

andersk commented:
Switching back and forth between an empty narrow and a nonempty narrow a couple of times gave me this error:

```
Uncaught TypeError: Cannot read property 'id' of undefined
    at Object../static/js/narrow.js.exports.maybe_add_local_messages (narrow.js:523)
    at Object../static/js/narrow.js.exports.activate (narrow.js:241)
    at Object../static/js/narrow.js.exports.by (narrow.js:683)
    at HTMLDivElement.<anonymous> (stream_list.js:490)
    at HTMLUListElement.dispatch (jquery.js:5429)
    at HTMLUListElement.elemData.handle (jquery.js:5233)
Uncaught TypeError: Cannot read property 'id' of undefined
    at Object../static/js/narrow.js.exports.maybe_add_local_messages (narrow.js:523)
    at Object../static/js/narrow.js.exports.activate (narrow.js:241)
    at do_hashchange_normal (hashchange.js:118)
    at hashchanged (hashchange.js:262)
    at hashchange.js:308
    at dispatch (jquery.js:5429)
    at elemData.handle (jquery.js:5233)
```

showell commented:
I can try to help with node tests once we're ready for those.  Let's try to coordinate that on chat when the time comes.

showell commented on static/js/narrow.js:
```diff
@@ -228,13 +231,20 @@ exports.activate = function (raw_operators, opts) {
         muting_enabled,
     });
 
+    const key = hash_util.operators_to_hash(operators);
```
Given that `MessageListDataCache` already knows how to use `hash_util`, it seems like we could tweak its interface here so that `narrow.js` doesn't have to understand this particular detail?

More broadly, this section of code just feels kinda clumsy, as we are first trying to create a `MessageListData` as `msg_data`, but then **after** that, we try to get an MLD from the cache, and we overwrite `msg_data` then.

showell commented:
I don't know exactly what the remedy here is, but we can probably simplify dependencies somewhat here among these things:

* MessageListDataCache (MLDC)
* narrow
* filter
* hash_util

It makes sense that MLDC would use filter and hash_util to build up its keys, but then we don't really fully encapsulate that, as my inline comments in `narrow.js` mention, since `narrow` still uses `hash_util`.  Since `narrow` already knows about `filter`, I almost wonder if the way to query the cache should be by passing in the `Filter` object?

ryanreh99 commented:
> Switching back and forth between an empty narrow and a nonempty narrow a couple of times gave me this error:

Thanks, I updated this PR to no longer cache empty narrows.



@showell could you have a look. I cleaned up the `MessageListDataCache` API.
Now, externally only a `filter` object needs to be passed (except for the `add` method).

The 3rd commits adds the tests, It is a bit large mainly due to the event fixtures.

The `message_events_caching.js` file tests whether the MLDCache contains the correct MLD object and each MLD object's messages correctly gets updated or not for each event.

timabbott commented:
Following https://github.com/zulip/zulip/pull/17883, I was reminded of this PR.  @ryanreh99 when you have a chance, it'd be really nice to rebase past the various relevant migrations -- I think we probably won't get this in for Zulip 4.0, but it'd be a really nice highlight for Zulip 5.0.

timabbott commented:
@ryanreh99 OK, now that 4.0 is released, I'd like to pick this up again; do you want to push on it a bit, perhaps in collaboration with @abhijeetbodas2001?

timabbott commented on static/js/message_list_data_cache.js:
```diff
@@ -0,0 +1,84 @@
+        // Internally, we serialize it, so this function
+        // should be called when only we want to access or
+        // create a stored MLD object.
+        return hash_util.operators_to_hash(filter.operators());
```
I think this algorithm isn't deterministic, in that if you have the same operators in a different order, it won't work.

timabbott replied:
(Or if I'm wrong and it is guaranteed to be sorted, we should have a comment explaining why, either here or in the `filter.operators()` definition)

timabbott commented on static/js/message_list_data_cache.js:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    get_valid_mlds(filter) {
+        const operators = filter.operators();
```
Can you add a comment explaining what policy this is trying to implement?  I wonder if we want something like "is_superset(filter1, filter2)" as the core logic bit of it.

timabbott commented on static/js/narrow.js:
```diff
@@ -269,17 +270,23 @@ export function activate(raw_operators, opts) {
+    if (mld_exists) {
+        msg_data = mld_cache.get(filter);
+    }
+
```
I think it might make sense to split this into first a commit that just implements the `MessageListDataCache` data structure, and adds test for it, and then adds a commit to store entries, and then proceeds with logic for actually using the cache.  

Part of my reasoning is that I think the logic for when to use the cache may end up being somewhat complex, potentially involving comparing the range of messages we have available from different data sources, and considering questions like whether the contiguous region that we have actually borders the message we need to land at.  And that will all require some pretty thoughtful testing, and potentially a series of smaller changes (E.g. at first we might want to only use the MessageListData cache when we know it includes the target message for the narrow).

timabbott commented on static/js/narrow.js:
```diff
@@ -269,17 +270,23 @@ export function activate(raw_operators, opts) {
+    const mld_exists = mld_cache.has(filter);
+    if (mld_exists) {
+        msg_data = mld_cache.get(filter);
+    }
```
This should probably be an if/else situation, where we don't call `new MessageListData` at all if we're using the cache?  Not sure, because of the discussion below.

timabbott commented on static/js/message_list_data_cache.js:
```diff
@@ -0,0 +1,84 @@
+            }
+        } else {
+            if (keys[0].includes("topic")) {
+                const stream_operator = [filter.operators()[0]];
```
I'm not sure this is correct; I think you want a `.operator("stream")` rather than assuming a specific ordering.

timabbott commented:
OK, I did an initial review on this, and posted comments on how we can break out a preparatory piece of this that should be mergeable as a partial checkpoint.  It might be worth doing a few more preparatory refactoring commits just to clean up things we'll want to end up changing more broadly; e.g. possibly we want to have the `filter` class expose a `.hash_key()` property as a prep commit, with tests, and then have this module just use that function, rather than having that be a feature of `MessageListDataCache`.  (I can imagine other data code wanting that, and in any case it seems like a cleaner abstraction boundary, and `filter.js` has a really nice unit test system).

I'd also like to add that we try hard to avoid abbreviations in the codebase, because they can be hard to read, so we should try to avoid writing a bunch of code using an `mld_cache` variable name.

timabbott commented:
One other note: We should think about how we're going to get to confidence that the new caching logic is 100% correct.  E.g. maybe we want to at first merge it with all the code to track the data structures in place, but in an "audit" mode, where effectively we compare the cache against what we get from fetching data from the server, and send a blueslip error when they disagree.
